kratos/python_scripts/multistage_analysis.py

```diff
@@ -0,0 +1,181 @@
+            # If not provided, create an auxiliary execution list from the stages declaration order
+            else:
+                KratosMultiphysics.Logger.PrintInfo("'execution_list' is not provided. Stages will be executed according to their declaration order.")
+                self.__execution_list = []
+                for stage_name in self.settings["stages"].keys():
+                    self.__execution_list.append(stage_name)
+
+        return self.__execution_list
+
+    def __GetModelers(self):
+        '''This method creates the modelers at the preprocess execution point.'''
+
+        #TODO: Add error thrown for wrong execution_point argument
+        execution_point_settings = self.settings["stages"][self.GetCurrentStageName()]["stage_preprocess"]
+
+        factory = KratosModelParametersFactory(self.model)
+        return factory.ConstructListOfItems(execution_point_settings["modelers"])
+
+    def __GetOperations(self, execution_point):
+        '''This method creates the operations at any execution point.'''
+
+        #TODO: Add error thrown for wrong execution_point argument
+        execution_point_settings = self.settings["stages"][self.GetCurrentStageName()][execution_point]
+
+        factory = KratosModelParametersFactory(self.model)
+        return factory.ConstructListOfItems(execution_point_settings["operations"])
+
+    def __CreateStagesMap(self):
+        '''This method creates the stages map.
+        The keys of the map are the stage names, which are retrieved from the execution list (see __GetExecutionList()).
+        The values of the map are the stages instances, which are created by importing its corresponding analysis stage Python module.
+        '''
+
+        stages_map = {}
+        for stage_name in self.__GetExecutionList():
+            analysis_stage_module_name = self.settings["stages"][stage_name]["analysis_stage"].GetString()
+            analysis_stage_class_name = analysis_stage_module_name.split('.')[-1]
+            analysis_stage_class_name = ''.join(x.title() for x in analysis_stage_class_name.split('_'))
+
+            analysis_stage_module = importlib.import_module(analysis_stage_module_name)
+            analysis_stage_class = getattr(analysis_stage_module, analysis_stage_class_name)
+
+            stages_map[stage_name] = self.__CreateAnalysisStageWithFlushInstance(analysis_stage_class, self.model,  KratosMultiphysics.Parameters(self.settings["stages"][stage_name]))
+        return stages_map
+
+    #TODO: Think about this
+    #TODO: I think we should be able to import a custom module in the current work directory
+    def __CreateAnalysisStageWithFlushInstance(self, cls, global_model, parameters):
+        '''This method emulates the old single-stage MainKratos.py.
+        Note that it exploits dynammic inheritance to extend the corresponding analysis stage class.
+        '''
+
+        class AnalysisStageWithFlush(cls):
```

Comment:
Can you please put this into a separate PR?
I think this should be integrated with the baseclass, I have somme comments/ideas (but lets not clutter this important PR) :)

Reply:
So you mean creating the stage as it is (without the flush) in this PR and then extending it?

Reply:
Removed :wink: .